### PR TITLE
Low power management module for PIC32

### PR DESCRIPTION
**Pull request by ghost**
This connects to #29.

Implement power management for I2C, SPI and UART.

### CODE REVIEW

File: platform/mikro-e/contiki-mikro-e-main.c
Comment (Avinash95) [inline platform/mikro-e/contiki-mikro-e-main.c]: void ?

File: cpu/pic32/lib/pic32_i2c.h
Comment (mtusnio) [inline cpu/pic32/lib/pic32_i2c.h]: By default all driver functions always return something, even if it's always 0.

File: cpu/pic32/lib/pic32_uart.h
Comment (mtusnio) [inline cpu/pic32/lib/pic32_uart.h]: Add a return type

**Review by mtusnio — CHANGES_REQUESTED** [review]

File: cpu/pic32/lib/pic32_i2c.c
Comment (Avinash95) [inline cpu/pic32/lib/pic32_i2c.c:78]: there is no error handling in the function, should return type be void ?

File: cpu/pic32/lpm.c
Comment (Avinash95) [inline cpu/pic32/lpm.c:83]: is this instruction aimed at to put the CPU in the idle mode ?

**Review by Avinash95 — DISMISSED** [review]
we are using timer1 and timer23, should we disable them as well ?

File: cpu/pic32/lpm.c
Comment (ghost) [reply cpu/pic32/lpm.c:83]: Yes, idle or sleep mode, it depends on the value of SLPEN bit.

**Comment by francois-berder-imgtec** [comment]
@Avinash95 timer2/3 are used for rtimer, and we need it to be able to wake-up the cpu in order to execute a task at a regular interval. Same for timer1 controlling the etimer.

File: cpu/pic32/lib/pic32_i2c.c
Comment (ghost) [reply cpu/pic32/lib/pic32_i2c.c:78]: No, I think it should still return ```int8_t``` to avoid changing the API in case we need to handle an error.

**Review by mtusnio — CHANGES_REQUESTED** [review]
SPI doesn't work for LED Matrix

**Comment by ghost** [comment]
I added some code to make LPM optional.

**Comment by ghost** [comment]
This pr is blocked on #69 .

**Comment by ghost** [comment]
@mtusnio It appears that disabling spi module used by cascoda and uart module used for debug output make the clicker impossible to wake-up. I think the clicker is probably stuck in an infinite loop.
As a consequence, I changed my code to enable LPM per module.

**Review by mtusnio — CHANGES_REQUESTED** [review]
SPI mode changes between power down/power up, UART does not work.

**Review by mtusnio — APPROVED** [review]

**Comment by mtusnio** [comment]
@Avinash95 can you review and approve if you are happy with the PR?

**Pull request merged**
